text/0000-parallel_ibd.md

```diff
@@ -0,0 +1,137 @@
+
+As such, a chunk merkle proof is simply a list of hashes:
+- Hashes of siblings along the path of subtree root to its peak (in case of a full chunk)
+- Hashes of the peaks in the MMR, from right to left, excluding the peaks that can be calculated given the chunk contents. For efficiency reasons the peaks on the right side of our subtree are bagged together
+
+### Sync process
+After receiving an output chunk, it should be verified according to the following steps:
+
+**Output chunk verification**
+- Verify `N_output == cardinality(bitmap)`
+- Fill unpruned leaves of output mmr subtree, calculate intermediary hashes where possible
+- Fill missing intermediary hashes from the chunk data and calculate final missing hashes
+- Use output chunk merkle proof to reconstruct `output_root`
+- Use bitmap merkle proof to reconstruct `bitmap_root`
+- Calculate `H(output_root|bitmap_root)` and compare with value in header
+- Verify rangeproofs
+
+If verification for this chunk passes, its contents should be stored somewhere. All unspent output commitments across all chunks need to eventually be summed up to a single commitment. The full output PMMR and bitmap MMR should be reconstructed as well. At which point in this process the (P)MMRs are constructed and the sum is calculated is left as an implementation choice.
+   
+The verification of kernel chunks is somewhat similar to output chunks, albeit somewhat simplified since kernels are never pruned:
+
+**Kernel chunk verification**
+- Fill leaves of kernel mmr subtree
+- Use kernel chunk merkle proof to reconstruct `kernel_root` and compare with value in header
+- Verify kernel signatures
+
+If verification of a kernel chunk is successful, it should be stored. Similar to the output chunk, we need to eventually calculate a single sum across all chunks. In this case it is a sum of all kernel excess commitments. The subtree data should also be used to reconstruct the full kernel MMR. Another wrinkle is that we need to "replay" the kernel root at each height to prevent certain attacks. Whether to update these quantities when each chunk is accepted or only calculate them at once after all chunks have been received is once again left open as an implementation choice.
+
+### Final check
+Once all output and kernel chunks are downloaded and verified, the (P)MMRs are recreated and the appropriate sums have been calculated we need to perform a final check:
+```
+output_sum =?= kernel_sum + TOTAL_SUPPLY * H
+```
+Here `TOTAL_SUPPLY` is the total supply of Grins at the point of sync, which is equal to `(HEIGHT + 1)*COINBASE_REWARD`.
+
+If all of this is successful, the node now has downloaded and verified the state up until the sync horizon point. Fully updating the chain state to the latest head involves downloading the full blocks after the sync horizon, verifying them and applying its contents consecutively, which is possible because all nodes are expected to store the full blocks past the compaction horizon.
+
```

Comment:
What happens when the checks fail? Process is restarted? Different peers? What's the peer selection process?

Reply:
There are a bunch of implementation details that I have intentionally left open, because I haven't fully decided on the scope of the RFC yet. Up to now I have limited it to mostly the p2p messages and how they should be verified, which is something we can comfortably get into the next hard fork release. But I could extend it to fully describe the new sync process, including how intermediary state should be tracked, how we can deal with malicious peers and what the plan would be around running the parallel sync side by side with the old sync at first, before we completely remove the old sync.